server/odc-service/src/main/java/com/oceanbase/odc/service/collaboration/environment/EnvironmentService.java

```diff
@@ -102,29 +148,89 @@ public Environment detailSkipPermissionCheck(@NonNull Long id) {
 
     @PreAuthenticate(actions = "update", resourceType = "ODC_ENVIRONMENT", indexOfIdParam = 0)
     @Transactional(rollbackFor = Exception.class)
-    public Environment update(@NonNull Long id, @NonNull Environment environment) {
-        EnvironmentEntity previous = environmentRepository.findByIdAndOrganizationId(id,
-                authenticationFacade.currentOrganizationId())
-                .orElseThrow(() -> new NotFoundException(ResourceType.ODC_ENVIRONMENT, "id", id));
+    public Environment update(@NotNull Long id, @NotNull UpdateEnvironmentReq req) {
+        Environment environment = innerDetail(id);
+        if (environment.getBuiltIn()) {
+            throw new BadRequestException("Not allowed to update builtin environments");
+        }
+        environment.setDescription(req.getDescription());
+        environment.setStyle(req.getStyle());
+        environment.setLastModifier(InnerUser.of(authenticationFacade.currentUserId(), null, null));
+        return entityToModel(environmentRepository.save(environmentMapper.modelToEntity(environment)));
+    }
 
-        previous.setDescription(environment.getDescription());
+    @PreAuthenticate(actions = "update", resourceType = "ODC_ENVIRONMENT", indexOfIdParam = 0)
+    @Transactional(rollbackFor = Exception.class)
+    public Boolean setEnabled(@NotNull Long id, @NotNull SetEnabledReq req) {
         /**
-         * TODO: permission check for ruleset
+         * ensure that the environment is not referenced by any data source before disabling it
          */
-        previous.setRulesetId(environment.getRulesetId());
-        previous.setLastModifierId(authenticationFacade.currentUserId());
+        if (!req.getEnabled()) {
+            checkDataSourceReference(id);
+        }
+        return environmentRepository.updateEnabledById(id, req.getEnabled()) > 0;
+    }
+
+    @PreAuthenticate(actions = "delete", resourceType = "ODC_ENVIRONMENT", indexOfIdParam = 0)
+    @Transactional(rollbackFor = Exception.class)
+    public Environment delete(@NotNull Long id) {
+        checkDataSourceReference(id);
+        checkRiskLevelReference(id);
+        Environment environment = innerDetail(id);
+        if (environment.getBuiltIn()) {
+            throw new BadRequestException("Not allowed to delete builtin environments");
+        }
+        environmentRepository.deleteById(id);
+        rulesetService.delete(environment.getRulesetId());
+        return environment;
+    }
+
+
+    private void checkDataSourceReference(Long id) {
+        Set<ConnectionConfig> referencedConnections =
+                connectionService.listByOrganizationId(authenticationFacade.currentOrganizationId()).stream()
```

Comment:
Why not list by environment id directly? May better performance.

Reply:
You are right! Updated.